sources/osg/Program.js

```diff
@@ -263,105 +271,170 @@ utils.createPrototypeStateAttribute(
                 var program = gl.createProgram();
                 this._vertex.failSafe(gl, this._vertex.getText());
                 this._fragment.failSafe(gl, this._fragment.getText());
-
-                gl.attachShader(program, this._vertex.shader);
-                gl.attachShader(program, this._fragment.shader);
-                gl.linkProgram(program);
-                gl.validateProgram(program);
-
+                this._glAttachAndCompile(gl, program, this._vertex, this._fragment);
                 notify.warn('FailSafe shader Activated ');
                 this._program = program;
             },
 
-            apply: function(state) {
-                if (this._nullProgram) return;
+            _glAttachAndCompile: function(gl, programGL, vertexShader, fragmentShader) {
+                gl.attachShader(programGL, vertexShader.shader);
+                gl.attachShader(programGL, fragmentShader.shader);
+                utils.timeStamp('osgjs.metrics:linkShader');
+                gl.linkProgram(programGL);
+            },
 
-                if (!this._gl) {
-                    this.setGraphicContext(state.getGraphicContext());
+            timeAndStart: function(msg) {
+                var current = Timer.instance().tick() - this._start;
+                if (current > limitSec) {
+                    console.log(msg + this._shaderName + ': ' + current + ' ms');
                 }
+                this._Accum += current;
+                this._start = Timer.instance().tick();
+            },
+            compile: function(state) {
                 var gl = this._gl;
-                if (!this._program || this._dirty) {
-                    var compileClean;
 
-                    if (!this._vertex.shader) {
-                        compileClean = this._vertex.compile(gl, errorCallback);
-                    }
+                var shaderName = this._fragment
+                    .getText()
+                    .match(/#define[\s]+SHADER_NAME[\s]+([\S]+)(\n|$)/);
 
-                    if (!this._fragment.shader) {
-                        compileClean = this._fragment.compile(gl, errorCallback);
-                    }
+                var compileClean = true;
 
-                    // do cpu operation before getting async compilation results
-                    var attributeMap = getAttributeList(this._vertex.getText());
-                    var attributeArray = window.Object.keys(attributeMap);
-                    var uniformMap = getUniformList(this._vertex.text + '\n' + this._fragment.text);
-                    var uniformArray = window.Object.keys(uniformMap);
+                this._shaderName = shaderName ? shaderName[1] : '';
 
-                    if (compileClean) {
-                        this._program = gl.createProgram();
+                this._start = Timer.instance().tick();
+                this._Accum = 0.0;
```

Comment:
Better to avoid variable names starting by a capital letter. Also declare it in the constructor.